### ISSUE TITLE
Filesystem dataset fails to mount until user login event

### ISSUE THREAD

**Issue by jpbrown-15** [zfsonlinux/zfs#9086]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| Ubuntu
Distribution Version	| 19.04
Linux Kernel	| 5.0.0-21-generic
Architecture	| x86_64
ZFS Version	| 0.7.12-1ubuntu5
SPL Version	| 0.7.12-1ubuntu3

### Describe the problem you're observing
I have a cron.daily job set to conduct snapshots of all zfs datasets and send them to a backup device. When I start the desktop without logging in (i.e. leaving the user selection screen displayed but not taking any action to select the user), the cron.daily job fails with "/usr/local/bin/pyznap: not found".  A login event, either directly on the desktop or via ssh and an ls of /usr/local/bin displays the correct content and the filesystem shows mounted.  This only occurs when a user does not login.  For instance, if I login as soon as the user selection screen appears, /usr/local/bin is mounted and the cron.daily job successfully completes.

### Describe how to reproduce the problem

1. Set up zfs on the root filesystem according to the Ubuntu 18.04 Root on ZFS instructions and include the optional rpool/usr and rpool/usr/local dataset.
2. Create a cron.daily job that invokes a command in /usr/local
3. Reboot the system and let it sit at the user selection screen (do not log in)
4. Wait 10 minutes, giving cron.daily time to determine it should run the daily job
5. Login and inspect /usr/local.  You should find that it is mounted and contains the content you expected.
6. Inspect the execution logs of cron.daily. You should see that the job failed due to not finding the executable within /usr/local.
7. Manually execute the cron.daily job to see that it successfully executes with a user logged in.

### Include any warning/errors/backtraces from the system logs

Original commands used to establish the /usr/local dataset on rpool:

```
History for 'rpool':
2019-02-23.07:05:36 zpool create -o ashift=12 -O canmount=off -O relatime=on -O compression=lz4 -O xattr=sa -O mountpoint=/ -R /mnt rpool /dev/disk/by-id/wwn-0x5000cca24ce19a84-part1
2019-02-23.07:06:46 zfs create -o canmount=off -o mountpoint=none rpool/ROOT
2019-02-23.07:07:22 zfs create -o canmount=noauto -o mountpoint=/ rpool/ROOT/ubuntu
2019-02-23.07:08:17 zfs create -o setuid=off rpool/home
2019-02-23.07:08:27 zfs create -o mountpoint=/root rpool/home/root
2019-02-23.07:08:37 zfs create -o canmount=off -o setuid=off -o exec=off rpool/var
2019-02-23.07:08:46 zfs create -o com.sun:auto-snapshot=false rpool/var/cache
2019-02-23.07:08:55 zfs create -o acltype=posixacl -o xattr=sa rpool/var/log
2019-02-23.07:09:03 zfs create rpool/var/spool
2019-02-23.07:09:13 zfs create -o com.sun:auto-snapshot=false -o exec=on rpool/var/tmp
2019-02-23.07:09:21 zfs create rpool/opt
2019-02-23.07:09:30 zfs create -o canmount=off rpool/usr
2019-02-23.07:09:37 zfs create rpool/usr/local
2019-02-23.07:09:47 zfs create rpool/var/mail
2019-02-23.07:09:56 zfs create -o exec=on rpool/var/spool/postfix
2019-02-23.07:10:06 zfs create -o com.sun:auto-snapshot=false -o setuid=off rpool/tmp
2019-02-23.10:28:34 zfs set mountpoint=legacy rpool/var/log
2019-02-23.10:28:47 zfs set mountpoint=legacy rpool/var/tmp
2019-02-23.10:29:10 zfs set mountpoint=legacy rpool/tmp
2019-02-23.10:49:51 zfs snapshot rpool/ROOT/ubuntu@install
```

Today's import at boot time:
```
2019-07-28.09:44:58 zpool import -N rpool

```
cron.daily ran at 9:51 and failed with:
```
/etc/cron.daily/zfs-backup:
/etc/cron.daily/zfs-backup: 16: /etc/cron.daily/zfs-backup: /usr/local/bin/pyznap: not found
/etc/cron.daily/zfs-backup: 21: /etc/cron.daily/zfs-backup: /usr/local/bin/pyznap: not found
run-parts: /etc/cron.daily/zfs-backup exited with return code 127
```

User login at 9:57 and conducted ls of /usr/local to see that file system was mounted.
```
jpb@explorer:~$ ls -l /usr/local/bin
total 375
lrwxrwxrwx 1 root root     26 Dec 19  2016 Moneydance -> /opt/Moneydance/Moneydance
-rwxr-xr-x 1 root root    225 Apr 22 10:09 pyznap
-rwxr-xr-x 1 root root    374 Feb 24 10:06 rsync-efi.sh
-rwxr-xr-x 1 root root    385 Mar 27  2018 snapper-gui
-rwxr-xr-x 1 root root 610296 Sep 24  2017 urbackupclientctl
-rwxr-xr-x 1 root root    299 Sep 28  2018 vm-time-sync.sh

```

Dmesg - no errors:
```
[    0.075818] Kernel command line: BOOT_IMAGE=/ROOT/ubuntu@/boot/vmlinuz-5.0.0-21-generic root=ZFS=rpool/ROOT/ubuntu ro intel_iommu=on i
ommu=pt quiet splash intel_iommu=on iommu=pt rootdelay=3 vt.handoff=1

[   40.298026] ZFS: Loaded module v0.7.12-1ubuntu5, ZFS pool version 5000, ZFS filesystem version 5

```

No errors in syslog.

Error in zfs-mount.service - /var/cache not empty:

```
jpb@explorer:~$ sudo systemctl status zfs-mount.service
[sudo] password for jpb: 
● zfs-mount.service - Mount ZFS filesystems
   Loaded: loaded (/lib/systemd/system/zfs-mount.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Sun 2019-07-28 09:45:19 CDT; 1h 16min ago
     Docs: man:zfs(8)
  Process: 3312 ExecStart=/sbin/zfs mount -a (code=exited, status=1/FAILURE)
 Main PID: 3312 (code=exited, status=1/FAILURE)

Jul 28 09:45:17 explorer systemd[1]: Starting Mount ZFS filesystems...
Jul 28 09:45:19 explorer zfs[3312]: cannot mount '/var/cache': directory is not empty
Jul 28 09:45:19 explorer systemd[1]: zfs-mount.service: Main process exited, code=exited, status=1/FAILURE
Jul 28 09:45:19 explorer systemd[1]: zfs-mount.service: Failed with result 'exit-code'.
Jul 28 09:45:19 explorer systemd[1]: Failed to start Mount ZFS filesystems.

```

**Comment by ttelford** [comment]
While the rest of the issue doesn't seem to apply to me, I get a similar `/var/cache/` error upon boot.  (Debian Sid, ZFS/SPL 0.8.1-3 -- though the error also happened in 0.7.12)

In my case, either `/var/cache/lxc` is mounted before `/var/cache`, or a dæmon loads (and puts files in `/var/cache`) before `/var/cache` is mounted by ZFS. (And in my case, `/var/cache/lxc` is mounted after `/var/cache` fails to mount.)

**Comment by jpbrown-15** [comment]
I cleaned up my /var/cache error. Somewhere between February and July, /var/cache didn't mount quick enough and got data in it.  I switched it to be a legacy mount which resolved the error.  It didn't however fix my main problem which consistently didn't mount /usr/local until a user login event.

This morning I changed the mountpoint to be legacy on /usr/local and placed it in fstab.  I'll see if that resolves the issue and report back.

It seems interesting that both rpool/usr and rpool/var are set to canmount=off and deeper datasets (like /var/log, /var/tmp, /var/cache, and /usr/local) that are set to canmount=on and have trouble.  It might be related to the race condition [#5754](https://github.com/zfsonlinux/zfs/issues/5754) -- at least /var/cache -- but the observation that /usr/local does get mounted after login is just strange.

**Comment by jpbrown-15** [comment]
Update. After changing /usr/local to legacy, syslog clearly stated /usr/local was mounted during the boot process and 5 minutes before cron.daily ran.  Yet, my cron.daily job failed as before.  Inspecting the the executable in /usr/local/bin that cron.daily invokes, it had a shebang that pointed back to my home directory which isn't mounted until I log in. That explains the observed behavior.  I moved the python component out of my home directory to /usr/local and corrected the shebang.  I expect all to work.

I've reverted rpool/usr/local to zfs mount control.

Sorry for the distraction.
